dnf/transaction_sr.py

```diff
@@ -0,0 +1,600 @@
+
+        swdb_group = self._base.history.group.new(group_id, comps_group.name, comps_group.ui_name, pkg_types)
+
+        try:
+            for pkg in pkgs:
+                name = pkg["name"]
+                self._assert_type(name, str, "groups.packages.name", "string")
+                installed = pkg["installed"]
+                self._assert_type(installed, bool, "groups.packages.installed", "boolean")
+                package_type = pkg["package_type"]
+                self._assert_type(package_type, str, "groups.packages.package_type", "string")
+
+                swdb_group.addPackage(name, installed, libdnf.transaction.stringToCompsPackageType(package_type))
+        except KeyError as e:
+            raise TransactionError(
+                _('Missing object key "{key}" in groups.packages.').format(key=e.args[0])
+            )
+
+        return swdb_group
+
+    def _swdb_group_install(self, group_id, pkg_types, pkgs):
+        self._base.history.group.install(self._create_swdb_group(group_id, pkg_types, pkgs))
+
+    def _swdb_group_upgrade(self, group_id, pkg_types, pkgs):
+        if not self._base.history.group.get(group_id):
+            self._raise_or_warn(
+                self._ignore_installed,
+                TransactionError(_("Group id '%s' is not installed.") % group_id)
+            )
+
+        self._base.history.group.upgrade(self._create_swdb_group(group_id, pkg_types, pkgs))
+
+    def _swdb_group_remove(self, group_id, pkg_types, pkgs):
+        if not self._base.history.group.get(group_id):
+            self._raise_or_warn(
+                self._ignore_installed,
+                TransactionError(_("Group id '%s' is not installed.") % group_id)
+            )
+
+        self._base.history.group.remove(self._create_swdb_group(group_id, pkg_types, pkgs))
+
+    def _create_swdb_environment(self, env_id, pkg_types, groups):
+        comps_env = self._base.comps._environment_by_id(env_id)
+        if not comps_env:
+            self._raise_or_warn(
+                self._skip_unavailable,
+                TransactionError(_("Environment id '%s' is not available.") % env_id)
+            )
+
+        swdb_env = self._base.history.env.new(env_id, comps_env.name, comps_env.ui_name, pkg_types)
+
+        try:
+            for grp in groups:
+                id = grp["id"]
+                self._assert_type(id, str, "environments.groups.id", "string")
+                installed = grp["installed"]
+                self._assert_type(installed, bool, "environments.groups.installed", "boolean")
+                group_type = grp["group_type"]
+                self._assert_type(group_type, str, "environments.groups.group_type", "string")
+
+                group_type = libdnf.transaction.stringToCompsPackageType(group_type)
+                if group_type not in (
+                    libdnf.transaction.CompsPackageType_MANDATORY,
+                    libdnf.transaction.CompsPackageType_OPTIONAL
+                ):
```

Comment:
Why you add only support for CompsPackageType_MANDATORY and CompsPackageType_OPTIONAL? Why not default type?

Reply:
Based on code from here onwards: https://github.com/rpm-software-management/dnf/blob/5d0e0d36e93d30e91314d5e7c311ee7927658b42/dnf/comps.py#L647

It would appear only `MANDATORY` and `OPTIONAL` is supported as `group_type`. Also, if I look at the description of the options here: https://fedoraproject.org/wiki/How_to_use_and_edit_comps.xml_for_package_groups#Comps_element_-_package, the difference between `MANDATORY` and `DEFAULT` doesn't make sense in `group_type` context.

Reply:
Ok